cluster.c

```diff
@@ -936,70 +939,129 @@ RRStatus ShardGroupParse(RedisModuleCtx *ctx, RedisModuleString **argv, int argc
             goto error;
         }
+        argidx++;
 
         memcpy(sg->nodes[i].node_id, str, len);
         sg->nodes[i].node_id[len] = '\0';
 
-        str = RedisModule_StringPtrLen(argv[argidx++], &len);
+        str = RedisModule_StringPtrLen(argv[argidx], &len);
         if (!NodeAddrParse(str, len, &sg->nodes[i].addr)) {
+            LOG_ERROR("ShardGroupParse failed; argv[%d] failed to parse node", base_argv_idx + argidx);
             RedisModule_ReplyWithError(ctx, "ERR invalid shard group message (node address/port)");
             goto error;
         }
+        argidx++;
     }
 
-    return RR_OK;
+    return num_argv_entries;
 
 error:
-    ShardGroupTerm(sg);
-    return RR_ERROR;
+    return -1;
 }
 
 /* parses all the shardgroups in a replace call and validates them */
 RRStatus ShardGroupsParse(RedisModuleCtx *ctx, RedisModuleString **argv, int argc, RaftReq *req)
 {
     if (RedisModule_StringToLongLong(argv[0], &req->r.shardgroups_replace.len) != REDISMODULE_OK) {
+        LOG_ERROR("Invalid shard group message: argv[0] (number of shard groups)");
         RedisModule_ReplyWithError(ctx, "ERR invalid shard group message(1)");
         return RR_ERROR;
     }
     req->r.shardgroups_replace.shardgroups =
             RedisModule_Calloc(req->r.shardgroups_replace.len, sizeof(ShardGroup *));
     argv++;
     argc--;
+    int argv_idx = 1;
 
     int i;
     for(i = 0; i < req->r.shardgroups_replace.len; i++) {
         ShardGroup *sg = ShardGroupCreate();
         req->r.shardgroups_replace.shardgroups[i] = sg;
 
-        long long num_slots, num_nodes;
-        if (RedisModule_StringToLongLong(argv[1], &num_slots) != REDISMODULE_OK ) {
-            RedisModule_ReplyWithError(ctx, "ERR invalid shard group message (num slots)");
-            goto fail;
-        }
-        if (RedisModule_StringToLongLong(argv[2], &num_nodes) != REDISMODULE_OK) {
-            RedisModule_ReplyWithError(ctx, "ERR invalid shard group message (num nodes)");
-            goto fail;
-        }
-
-        int num_argv_entries = (int) (3 + num_slots * 3 + num_nodes * 2);
-        if (num_argv_entries > argc) {
-            RedisModule_ReplyWithError(ctx, "ERR invalid shard group message (num entries)");
-            goto fail;
-        }
-
-        if (ShardGroupParse(ctx, argv, num_argv_entries, sg) != RR_OK) {
+        int num_argv_entries = 0;
+        if ((num_argv_entries = ShardGroupParse(ctx, argv, argc, sg, argv_idx)) < 0) {
             goto fail;
         }
 
         argv += num_argv_entries;
         argc -= num_argv_entries;
+        argv_idx += num_argv_entries;
+    }
+
+    if (argc != 0) {
+        LOG_ERROR("Invalid shard group message: extra unexpected arguments");
+        RedisModule_ReplyWithError(ctx, "ERR invalid shard group message (extra arguments)");
+        goto fail;
+    }
+
+    /* basic validation, no shard groups have slotranges that overlap another */
+    ShardGroup * stable[REDIS_RAFT_HASH_MAX_SLOT+1];
```

Comment:
nit: pointer style (RedisRaft uses `ShardGroup *var` elsewhere).
Also, why not `ShardGroup *var[...] = { 0 }`?

Reply:
I guess I didn't intuitively know that one can do that, changed the code.